File: orchestra/channels.py
"""Small channels useful in examples and local operation."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Callable, Mapping

from .models import OK, HealthReport, TickReport
from .state import read_json, write_json_atomic


class ConsoleChannel:
    name = "console"

    def __init__(self):
        self._health = HealthReport(OK, "ready")

    def publish(self, report: TickReport) -> None:
        for action in report.actions:
            print(action)
        if not report.actions:
            print("no actions")
        self._health = HealthReport(OK, f"reported {len(report.actions)} action(s)")

    def health(self) -> HealthReport:
        return self._health


class JsonlAuditChannel:
    """Append one compact, content-free tick record to a local JSONL log."""

    name = "audit-log"

    def __init__(self, path: Path):
        self.path = Path(path)
        self._health = HealthReport(OK, "ready")

    def publish(self, report: TickReport) -> None:
        if report.dry_run:
            self._health = HealthReport(OK, "dry run; audit not written")
            return
        self.path.parent.mkdir(parents=True, exist_ok=True)
        row = {
            "started_at": report.started_at,
            "finished_at": report.finished_at,
            "dry_run": report.dry_run,
            "actions": report.actions,
            "states": {
                key: snapshot.state for key, snapshot in report.snapshots.items()
            },
            "health": {
                name: {"status": value.status, "message": value.message}
                for name, value in report.health.items()
            },
            "degraded_sources": report.degraded_sources,
        }
        with self.path.open("a", encoding="utf-8") as handle:
            handle.write(json.dumps(row, sort_keys=True) + "\n")
        self._health = HealthReport(OK, "tick appended")

    def health(self) -> HealthReport:
        return self._health


def health_changed(
    previous: Mapping[str, object], name: str, current: HealthReport
) -> bool:
    row = previous.get(name)
    if not isinstance(row, dict) or row.get("status") != current.status:
        return True
    return current.status != OK and row.get("message") != current.message


class HealthChangeChannel:
    """Report edge-triggered component health changes through any text sink."""

    name = "health"

    def __init__(self, state_path: Path, sink: Callable[[str], None] = print):
        self.state_path = Path(state_path)
        self.sink = sink
        self._health = HealthReport(OK, "ready")

    def publish(self, report: TickReport) -> None:
        if report.dry_run:
            self._health = HealthReport(OK, "dry run; health state not written")
            return
        previous = read_json(self.state_path) or {}
        lines = [
            f"{value.status} {name}: {value.message}"
            for name, value in sorted(report.health.items())
            if health_changed(previous, name, value)
        ]
        if lines:
            self.sink("\n".join(lines))
        write_json_atomic(
            self.state_path,
            {
                name: {"status": value.status, "message": value.message}
                for name, value in report.health.items()
            },
        )
        self._health = HealthReport(OK, f"reported {len(lines)} change(s)")

    def health(self) -> HealthReport:
        return self._health

File: orchestra/channels.ts
import { appendFileSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";
import { health, OK } from "./models.js";
import type { HealthReport, TickReport } from "./models.js";
import type { Awaitable, Channel } from "./ports.js";
import { readJson, writeJsonAtomic } from "./state.js";
import { isRecord } from "./json.js";

export class ConsoleChannel implements Channel {
  readonly name = "console";
  private current = health(OK, "ready");
  constructor(private readonly sink: (text: string) => void = console.log) {}
  publish(report: TickReport): void {
    for (const action of report.actions.length
      ? report.actions
      : ["no actions"])
      this.sink(action);
    this.current = health(OK, `reported ${report.actions.length} action(s)`);
  }
  health(): HealthReport {
    return this.current;
  }
}
export class JsonlAuditChannel implements Channel {
  readonly name = "audit-log";
  private current = health(OK, "ready");
  constructor(readonly path: string) {}
  publish(report: TickReport): void {
    if (report.dryRun) {
      this.current = health(OK, "dry run; audit not written");
      return;
    }
    mkdirSync(dirname(this.path), { recursive: true });
    appendFileSync(
      this.path,
      JSON.stringify({
        started_at: report.startedAt,
        finished_at: report.finishedAt,
        dry_run: report.dryRun,
        actions: report.actions,
        states: Object.fromEntries(
          Object.entries(report.snapshots).map(([key, value]) => [
            key,
            value.state,
          ]),
        ),
        health: report.health,
        degraded_sources: report.degradedSources,
      }) + "\n",
    );
    this.current = health(OK, "tick appended");
  }
  health(): HealthReport {
    return this.current;
  }
}
export function healthChanged(
  previous: Readonly<Record<string, unknown>>,
  name: string,
  current: HealthReport,
): boolean {
  const row = previous[name];
  return (
    !isRecord(row) ||
    row.status !== current.status ||
    (current.status !== OK && row.message !== current.message)
  );
}
export class HealthChangeChannel implements Channel {
  readonly name = "health";
  private current = health(OK, "ready");
  constructor(
    readonly statePath: string,
    private readonly sink: (text: string) => Awaitable<void> = console.log,
  ) {}
  async publish(report: TickReport): Promise<void> {
    if (report.dryRun) {
      this.current = health(OK, "dry run; health state not written");
      return;
    }
    const previous = readJson(this.statePath) ?? {};
    const lines = Object.entries(report.health)
      .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
      .filter(([name, value]) => healthChanged(previous, name, value))
      .map(([name, value]) => `${value.status} ${name}: ${value.message}`);
    if (lines.length) await this.sink(lines.join("\n"));
    writeJsonAtomic(this.statePath, report.health);
    this.current = health(OK, `reported ${lines.length} change(s)`);
  }
  health(): HealthReport {
    return this.current;
  }
}
